packages/next-codemod/transforms/next-request-geo-ip.ts

```diff
@@ -0,0 +1,451 @@
+        tsIndexedAccessType.objectType.typeName.type === 'Identifier' &&
+        tsIndexedAccessType.objectType.typeName.name === 'NextRequest' &&
+        tsIndexedAccessType.indexType.type === 'TSLiteralType' &&
+        tsIndexedAccessType.indexType.literal.type === 'StringLiteral' &&
+        tsIndexedAccessType.indexType.literal.value === IP
+      )
+    }
+  )
+
+  if (nextReqGeoType.length > 0) {
+    needImportGeoType = true
+  }
+
+  // replace with type Geo
+  nextReqGeoType.replaceWith(j.identifier(geoTypeIdentifier))
+  // replace with type string | undefined
+  nextReqIpType.replaceWith(
+    j.tsUnionType([j.tsStringKeyword(), j.tsUndefinedKeyword()])
+  )
+
+  return {
+    needImportGeoType,
+  }
+}
+
+function insertImportDeclarations(
+  j: API['jscodeshift'],
+  ast: Collection<Node>,
+  needImportGeolocation: boolean,
+  needImportIpAddress: boolean,
+  needImportGeoType: boolean,
+  geoIdentifier: string,
+  ipIdentifier: string,
+  geoTypeIdentifier: string
+) {
+  const firstNextServerImport = ast
+    .find(j.ImportDeclaration, {
+      source: {
+        value: 'next/server',
+      },
+    })
+    // get the first import declaration
+    .at(0)
+
+  if (needImportGeolocation || needImportIpAddress) {
+    const importDeclaration = j.importDeclaration(
+      [
+        needImportGeolocation
+          ? j.importSpecifier(
+              j.identifier(GEOLOCATION),
+              // If there was a duplicate identifier, we add an
+              // incremental number suffix to it and we use alias:
+              // `import { geolocation as geolocation1 } from ...`
+              j.identifier(geoIdentifier)
+            )
+          : null,
+        needImportIpAddress
+          ? j.importSpecifier(
+              j.identifier(IP_ADDRESS),
+              // If there was a duplicate identifier, we add an
+              // incremental number suffix to it and we use alias:
+              // `import { ipAddress as ipAddress1 } from ...`
+              j.identifier(ipIdentifier)
+            )
+          : null,
+      ].filter(Boolean),
+      j.literal('@vercel/functions')
+    )
+
+    firstNextServerImport.insertAfter(importDeclaration)
+  }
+
+  if (needImportGeoType) {
+    const geoTypeImportDeclaration = j.importDeclaration(
+      [
+        needImportGeoType
+          ? j.importSpecifier(
+              j.identifier(GEO_TYPE),
+              // If there was a duplicate identifier, we add an
+              // incremental number suffix to it and we use alias:
+              // `import { Geo as Geo1 } from ...`
+              j.identifier(geoTypeIdentifier)
+            )
+          : null,
+      ].filter(Boolean),
+      j.literal('@vercel/functions'),
+      // Not using inline type import because it violates:
+      // https://typescript-eslint.io/rules/no-import-type-side-effects
+      'type'
+    )
+    firstNextServerImport.insertAfter(geoTypeImportDeclaration)
```

Comment:
If vercel/functions import has presented. find the import of vercel/functions and push the type into the import specifiers.

Reply:
For the inline type, same reason with:

> It's to avoid inline type because it violates:
> x-ref: https://typescript-eslint.io/rules/no-import-type-side-effects
> x-ref: https://github.com/eps1lon/types-react-codemod/blob/main/transforms/scoped-jsx.js#L53-L54

Reply:
But we could do that for `geolocation` & `ipAddress`!

Reply:
Update: made it inline if there's a vercel func import:

```diff
- import { waitUntil } from '@vercel/functions'
+ import { waitUntil, type Geo } from '@vercel/functions'
```

or else make it type import.

```diff
+ import type { Geo } from '@vercel/functions'
```